File: src/wu/seal/jsontokotlin/MakeKotlinClassAction.java
package wu.seal.jsontokotlin;

import com.google.gson.*;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.command.CommandProcessor;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.InputValidator;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.util.ui.JBDimension;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.text.JTextComponent;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.util.IllegalFormatFlagsException;

/**
 * Created by Seal.Wu on 2017/8/18.
 */
public class MakeKotlinClassAction extends AnAction {
    // If you register the action from Java code, this constructor is used to set the menu item name
    // (optionally, you can specify the menu description and an icon to display next to the menu item).
    // You can omit this constructor when registering the action in the plugin.xml file.
    public MakeKotlinClassAction() {
        // Set the menu item name.
        super("MakeKotlinClass");
        // Set the menu item name, description and icon.
        // super("Text _Boxes","Item description",IconLoader.getIcon("/Mypackage/icon.png"));
    }

    public void actionPerformed(AnActionEvent event) {
        Project project = event.getData(PlatformDataKeys.PROJECT);
        final Caret caret = event.getData(PlatformDataKeys.CARET);
        final Editor editor = event.getData(PlatformDataKeys.EDITOR_EVEN_IF_INACTIVE);
        if (editor == null) {
            Messages.showWarningDialog("Please open a file in editor state for insert Kotlin code!", "No Editor File");
            return;
        }
        final String className = Messages.showInputDialog(project, "Please input the Class Name for Insert", "Input ClassName", Messages.getInformationIcon());
        if (className == null || className.isEmpty()) {
            return;
        }
        final Messages.InputDialog inputDialog = new Messages.InputDialog(project, "Please input the Json Data", "Input Json"
                , Messages.getInformationIcon(), "", new InputValidator() {
            @Override
            public boolean checkInput(String inputString) {
                try {
                    JsonElement jsonElement = new JsonParser().parse(inputString);
                    return jsonElement.isJsonObject() || jsonElement.isJsonArray();
                } catch (JsonSyntaxException e) {
                    return false;
                }
            }

            @Override
            public boolean canClose(String inputString) {
                return true;
            }
        }) {
            @NotNull
            protected JPanel createMessagePanel() {
                JPanel messagePanel = new JPanel(new BorderLayout());
                if (myMessage != null) {
                    JComponent textComponent = createTextComponent();
                    messagePanel.add(textComponent, BorderLayout.NORTH);
                }

                myField = createTextFieldComponent();


                messagePanel.add(createScrollableTextComponent(), BorderLayout.CENTER);
                JButton settingButton = new JButton("Config Settings");
                settingButton.addActionListener(new AbstractAction() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        new ConfigSettingDialog(false).show();
                    }
                });
                JPanel settingContainer = new JPanel();
                BoxLayout boxLayout = new BoxLayout(settingContainer, BoxLayout.LINE_AXIS);
                settingContainer.setLayout(boxLayout);
                settingButton.setHorizontalAlignment(SwingConstants.RIGHT);
                settingContainer.add(settingButton);
                messagePanel.add(settingContainer, BorderLayout.SOUTH);
                return messagePanel;
            }

            @Override
            protected JTextComponent createTextFieldComponent() {
                JTextArea jTextArea = new JTextArea(15, 100);
                jTextArea.setMinimumSize(new JBDimension(800, 500));
                jTextArea.setMaximumSize(new JBDimension(1000, 700));
                jTextArea.setLineWrap(true);
                jTextArea.setWrapStyleWord(true);
                jTextArea.setAutoscrolls(true);
                return jTextArea;
            }


            protected JComponent createScrollableTextComponent() {
                return new JBScrollPane(myField);
            }
        };
        inputDialog.show();
        String jsonString = inputDialog.getInputString();
        if (jsonString == null || jsonString.isEmpty()) {
            return;
        }
        final Document document = editor.getDocument();
        ImportClassWriter.INSTANCE.insertImportClassCode(project, document);

        final KotlinMaker maker;
        try {
            maker = new KotlinMaker(className, jsonString);
        } catch (IllegalFormatFlagsException e) {
            e.printStackTrace();
            Messages.showErrorDialog(e.getMessage(), "Unsupport Json");
            return;
        }

        CommandProcessor.getInstance().executeCommand(project, new Runnable() {
            @Override
            public void run() {
                ApplicationManager.getApplication().runWriteAction(new Runnable() {
                    @Override
                    public void run() {

                        int offset = 0;

                        if (caret != null) {

                            offset = caret.getOffset();
                            if (offset == 0) {
                                offset = document.getTextLength() - 1;
                            }
                        } else {
                            offset = document.getTextLength() - 1;
                        }
                        document.insertString(Math.max(offset, 0), maker.makeKotlinData());

                    }
                });
            }
        }, "insertKotlin", "JsonToKotlin");
        Messages.showMessageDialog(project, "Kotlin Code insert successfully!", "Information", Messages.getInformationIcon());
    }
}

File: src/wu/seal/jsontokotlin/MakeKotlinClassAction.kt
package wu.seal.jsontokotlin

import com.google.gson.Gson
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.editor.Caret
import com.intellij.openapi.editor.Document
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import wu.seal.jsontokotlin.feedback.StartAction
import wu.seal.jsontokotlin.feedback.SuccessCompleteAction
import wu.seal.jsontokotlin.feedback.getUncaughtExceptionHandler
import wu.seal.jsontokotlin.feedback.sendActionInfo
import wu.seal.jsontokotlin.ui.JsonInputDialog
import wu.seal.jsontokotlin.utils.executeCouldRollBackAction

import java.util.IllegalFormatFlagsException

/**
 * Plugin action
 * Created by Seal.Wu on 2017/8/18.
 */
class MakeKotlinClassAction : AnAction("MakeKotlinClass") {

    private val gson = Gson()

    override fun actionPerformed(event: AnActionEvent) {
        var jsonString: String = ""
        try {
            actionStart()
            val project = event.getData(PlatformDataKeys.PROJECT)
            val caret = event.getData(PlatformDataKeys.CARET)
            val editor = event.getData(PlatformDataKeys.EDITOR_EVEN_IF_INACTIVE)
            if (couldNotInsertCode(editor)) return
            val document = editor!!.document
            val editorText = document.text
            /**
             *  temp class name for insert
             */
            var tempClassName = ""
            val couldGetAndReuseClassNameInCurrentEditFileForInsertCode = couldGetAndReuseClassNameInCurrentEditFileForInsertCode(editorText)

            if (couldGetAndReuseClassNameInCurrentEditFileForInsertCode) {
                /**
                 * auto obtain the current class name
                 */
                tempClassName = getCurrentEditFileTemClassName(editorText)
            }
            val inputDialog = JsonInputDialog(tempClassName, project!!)
            inputDialog.show()
            val className = inputDialog.getClassName()
            val json = inputDialog.inputString
            if (json == null || json.isEmpty()) {
                return
            }
            jsonString = json

            if (reuseClassName(couldGetAndReuseClassNameInCurrentEditFileForInsertCode, className, tempClassName)) {
                executeCouldRollBackAction(project) {
                    /**
                     * if you don't clean then we will trick a conflict with two same class name error
                     */
                    cleanCurrentEditFile(document)
                }
            }
            if (insertKotlinCode(project, document, className, jsonString, caret)) {
                actionComplete()
            }


        } catch(e: Exception) {
            dealWithException(jsonString, e)
            throw e
        }
    }

    private fun reuseClassName(couldGetAndReuseClassNameInCurrentEditFileForInserCode: Boolean, className: String, tempClassName: String) = couldGetAndReuseClassNameInCurrentEditFileForInserCode && className == tempClassName

    private fun couldNotInsertCode(editor: Editor?): Boolean {
        if (editor == null) {
            Messages.showWarningDialog("Please open a file in editor state for insert Kotlin code!", "No Editor File")
            return true
        }
        return false
    }

    private fun actionComplete() {
        Thread {
            sendActionInfo(gson.toJson(SuccessCompleteAction()))
        }.start()
    }

    private fun actionStart() {
        Thread {
            sendActionInfo(gson.toJson(StartAction()))
        }.start()
    }

    private fun dealWithException(jsonString: String, e: Exception) {
        var jsonString1 = jsonString
        val yes = Messages.showYesNoDialog("Some thing execute wrong.\nAgree with publishing your JSON text to help us to solve the problem?", "Excuse me", Messages.getQuestionIcon())
        if (yes != Messages.YES) {
            jsonString1 = "User keep private about JSON text"
        }
        getUncaughtExceptionHandler(jsonString1) {
            Messages.showErrorDialog("I am sorry,JsonToKotlinClass may occur a RuntimeException,\nYou could try again later or recover to the old version,\nOr you could post an issue here:\nhttps://github.com/wuseal/JsonToKotlinClass\nWe will fixed it soon!", "Occur a fatal error")
        }.uncaughtException(Thread.currentThread(), e)
    }

    private fun insertKotlinCode(project: Project?, document: Document, className: String, jsonString: String, caret: Caret?): Boolean {
        ImportClassWriter.insertImportClassCode(project, document)

        val codeMaker: KotlinCodeMaker
        try {
            codeMaker = KotlinCodeMaker(className, jsonString)
        } catch (e: IllegalFormatFlagsException) {
            e.printStackTrace()
            Messages.showErrorDialog(e.message, "UnSupport Json")
            return false
        }

        executeCouldRollBackAction(project) {
            var offset: Int

            if (caret != null) {

                offset = caret.offset
                if (offset == 0) {
                    offset = document.textLength - 1
                }
            } else {
                offset = document.textLength - 1
            }
            document.insertString(Math.max(offset, 0), codeMaker.makeKotlinData())
        }
        return true
    }

    private fun cleanCurrentEditFile(document: Document, editorText: String = document.text) {
        document.setText(editorText.substringBefore("class"))
    }

    private fun getCurrentEditFileTemClassName(editorText: String) = editorText.substringAfter("class").substringBefore("{").trim()

    /**
     * whether we could reuse current class name declared in the edit file for inserting data class code
     * if we could use it,then we would clean the kotlin file as it was new file without any class code .
     */
    private fun couldGetAndReuseClassNameInCurrentEditFileForInsertCode(editorText: String): Boolean {
        var couldGetAndReuseClassNameInCurrentEditFileForInsertCode = false
        if (editorText.indexOf("class") == editorText.lastIndexOf("class")
                && editorText.substringAfter("class").contains("(").not()
                && editorText.substringAfter("class").contains(":").not()
                && editorText.substringAfter("class").contains("=").not()) {
            couldGetAndReuseClassNameInCurrentEditFileForInsertCode = true
        }
        return couldGetAndReuseClassNameInCurrentEditFileForInsertCode
    }
}
